trin-core/src/jsonrpc/handlers.rs

```diff
@@ -6,113 +6,78 @@ use serde_json::Value;
-pub struct HistoryNetworkEndpoint {
-    pub kind: HistoryEndpointKind,
-    pub resp: Responder<Value, String>,
-}
-
-#[derive(Debug)]
-pub enum StateEndpointKind {
-    GetStateNetworkData,
-}
-
-#[derive(Debug)]
-pub struct StateNetworkEndpoint {
-    pub kind: StateEndpointKind,
-    pub resp: Responder<Value, String>,
-}
-
-#[derive(Debug)]
-pub enum PortalEndpointKind {
-    DummyHistoryNetworkData,
-    DummyStateNetworkData,
-    NodeInfo,
-    RoutingTableInfo,
-}
-
-#[derive(Debug)]
-pub struct PortalEndpoint {
-    pub kind: PortalEndpointKind,
-    pub params: Params,
-    pub resp: Responder<Value, String>,
-}
-
+/// Main JSON-RPC handler. It dispatches json-rpc requests to the overlay networks.
 pub struct JsonRpcHandler {
     pub discovery: Arc<RwLock<Discovery>>,
-    pub jsonrpc_rx: mpsc::UnboundedReceiver<PortalEndpoint>,
-    pub state_tx: Option<mpsc::UnboundedSender<StateNetworkEndpoint>>,
-    pub history_tx: Option<mpsc::UnboundedSender<HistoryNetworkEndpoint>>,
+    pub portal_jsonrpc_rx: mpsc::UnboundedReceiver<PortalJsonRpcRequest>,
+    pub state_jsonrpc_tx: Option<mpsc::UnboundedSender<StateJsonRpcRequest>>,
+    pub history_jsonrpc_tx: Option<mpsc::UnboundedSender<HistoryJsonRpcRequest>>,
 }
 
 impl JsonRpcHandler {
     pub async fn process_jsonrpc_requests(mut self) {
-        while let Some(cmd) = self.jsonrpc_rx.recv().await {
-            use PortalEndpointKind::*;
-
-            match cmd.kind {
-                NodeInfo => {
-                    let node_id = self.discovery.read().await.local_enr().to_base64();
-                    let _ = cmd.resp.send(Ok(Value::String(node_id)));
-                }
-                RoutingTableInfo => {
-                    let routing_table_info = self
-                        .discovery
-                        .read()
-                        .await
-                        .discv5
-                        .table_entries_id()
-                        .iter()
-                        .map(|node_id| Value::String(node_id.to_string()))
-                        .collect();
-                    let _ = cmd.resp.send(Ok(Value::Array(routing_table_info)));
-                }
-                DummyHistoryNetworkData => {
-                    let response = match self.history_tx.as_ref() {
-                        Some(tx) => {
-                            proxy_query_to_history_subnet(
-                                tx,
-                                HistoryEndpointKind::GetHistoryNetworkData,
-                            )
+        while let Some(request) = self.portal_jsonrpc_rx.recv().await {
+            match request.endpoint {
+                PortalEndpointKind::Discv5EndpointKind(endpoint) => match endpoint {
+                    Discv5EndpointKind::NodeInfo => {
+                        let node_id = self.discovery.read().await.local_enr().to_base64();
+                        let _ = request.resp.send(Ok(Value::String(node_id)));
+                    }
```

Comment:
Hm, the list of `discv5_` things to do could get very long. My instinct is to minimize as much as possible by having each match branch be a one-liner. (Though maybe ) That could look something like:
```rs
            let response = match request.endpoint {
                PortalEndpointKind::Discv5EndpointKind(endpoint) => match endpoint {
                    Discv5EndpointKind::NodeInfo => self.discv5.node_info(),
                    # ...
                }
            }
            let _ = request.resp.send(response);
```

Where `discv5` is a class responsible for implementing all the `discv5_*` handlers, like `node_info()`, which does this:
```rs
let node_id = self.discovery.read().await.local_enr().to_base64();
Ok(Value::String(node_id))
```

I think that refactor translates nicely enough for the other calls that get proxied through to other handlers, too.

Reply:
No sure if we can move the response outside the `match` block, because It looks like some responses require different `serde_json::Value` wrappers (String, Array, Number). The handler has access to discv5 class (self.discovery), what we can do is to add helper methods in `self.discovery` for things like `node_id`.